bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableDataSettingsFactory.java

```diff
@@ -0,0 +1,310 @@
+
+    // TODO: implementation for channelCount or channelPerCPU
+    ManagedChannelBuilder channelBuilder = ManagedChannelBuilder
+        .forAddress(options.getDataHost(), options.getPort())//
+        .userAgent(options.getUserAgent());
+
+    if (options.usePlaintextNegotiation()) {
+      channelBuilder.usePlaintext();
+    }
+    builder.setTransportChannelProvider(
+      FixedTransportChannelProvider.create(GrpcTransportChannel.create(channelBuilder.build())));
+
+    return builder.build();
+  }
+
+  /**
+   * This method is use to build BatchSettings.
+   *
+   * @param builder a {@link BigtableDataSettings.Builder} object.
+   * @param options a {@link BigtableOptions} object.
+   */
+  private static void buildBulkOptions(Builder builder, BigtableOptions options) {
+    BulkOptions bulkOptions = options.getBulkOptions();
+    BatchingSettings.Builder batchSettingsBuilder = BatchingSettings.newBuilder();
+
+    // NOTE: FlowControlSettings by default uses LimitExceededBehavior.Block
+    FlowControlSettings.Builder flowControlBuilder =
+        FlowControlSettings.newBuilder()
+            .setMaxOutstandingRequestBytes(bulkOptions.getMaxMemory());
+
+    long autoFlushMs = bulkOptions.getAutoflushMs();
+    if (autoFlushMs > 0) {
+      batchSettingsBuilder.setDelayThreshold(Duration.ofMillis(autoFlushMs));
+    }
+    long maxIn = bulkOptions.getMaxInflightRpcs();
+    if(maxIn > 0) {
+      flowControlBuilder.setMaxOutstandingElementCount(maxIn);
+    }
+
+    batchSettingsBuilder
+        .setIsEnabled(bulkOptions.useBulkApi())
+        .setElementCountThreshold(Long.valueOf(bulkOptions.getBulkMaxRowKeyCount()))
+        .setRequestByteThreshold(bulkOptions.getBulkMaxRequestSize())
+        .setFlowControlSettings(flowControlBuilder.build());
+
+    // TODO: implement bulkMutationThrottling & bulkMutationRpcTargetMs, once available
+    if (options.getRetryOptions().enableRetries()) {
+      builder.bulkMutationsSettings()
+          .setBatchingSettings(batchSettingsBuilder.build())
+          .setRetryableCodes(DEFAULT_RETRY_CODES)
+          .setRetrySettings(defaultRetrySettings(options));
+    } else {
+      builder.bulkMutationsSettings()
+          .setSimpleTimeoutNoRetries(DEFAULT_RPC_TIMEOUT)
+          .getRetryableCodes().clear();
+    }
+  }
+
+  /**
+   * To build sampleRowKey with default Settings based on retries setting.
+   *
+   * @param builder a {@link BigtableDataSettings.Builder} object.
+   * @param bulkMutation a {@link BulkOptions} object.
+   */
+  private static void buildSampleRowKeys(Builder builder, BigtableOptions options) {
+    if (options.getRetryOptions().enableRetries()) {
+      builder.sampleRowKeysSettings()
+          .setRetryableCodes(DEFAULT_RETRY_CODES)
+          .setRetrySettings(defaultRetrySettings(options));
+    } else {
+      builder.sampleRowKeysSettings()
+          .setSimpleTimeoutNoRetries(DEFAULT_RPC_TIMEOUT)
+          .getRetryableCodes().clear();
```

Comment:
doesn't setSimpleTimeoutNoRetries already clear the codes?

Reply:
Yes, Will remove unnecessary `getRetryableCodes().clear()`